src/server/cluster/outgoing_slot_migration.cc

```diff
@@ -81,48 +88,108 @@ OutgoingMigration::~OutgoingMigration() {
   main_sync_fb_.JoinIfNeeded();
 }
 
-void OutgoingMigration::Finish() {
-  shard_set->pool()->AwaitFiberOnAll([this](util::ProactorBase* pb) {
-    if (const auto* shard = EngineShard::tlocal(); shard)
-      slot_migrations_[shard->shard_id()]->Cancel();
-  });
-  state_.store(MigrationState::C_FINISHED);
+void OutgoingMigration::Finish(bool is_error) {
+  std::lock_guard lk(finish_mu_);
+  if (state_.load() != MigrationState::C_FINISHED || state_.load() != MigrationState::C_ERROR) {
+    shard_set->pool()->AwaitFiberOnAll([this](util::ProactorBase* pb) {
+      if (const auto* shard = EngineShard::tlocal(); shard)
+        slot_migrations_[shard->shard_id()]->Cancel();
+    });
+    const auto new_state = is_error ? MigrationState::C_ERROR : MigrationState::C_FINISHED;
+    state_.store(new_state);
+  }
 }
 
 MigrationState OutgoingMigration::GetState() const {
   return state_.load();
 }
 
 void OutgoingMigration::SyncFb() {
-  state_.store(MigrationState::C_SYNC);
-  auto start_cb = [this](util::ProactorBase* pb) {
-    if (auto* shard = EngineShard::tlocal(); shard) {
-      server_family_->journal()->StartInThread();
-      slot_migrations_[shard->shard_id()] = std::make_unique<SliceSlotMigration>(
-          &shard->db_slice(), server(), migration_info_.slot_ranges, server_family_->journal(),
-          &cntx_);
-      slot_migrations_[shard->shard_id()]->Start(cf_->MyID(), shard->shard_id());
+  // we retry starting migration until "cancel" is happened
+  while (state_.load() != MigrationState::C_FINISHED) {
+    state_.store(MigrationState::C_CONNECTING);
+    last_error_ = cntx_.GetError();
+    cntx_.Reset(nullptr);
```

Comment:
There is `SwitchErrorHandler` to do this atomically, because if it's cancelled in-between those lines, you'll loose the error

Reply:
I need to reset error here, It's ok if I loose error during cancel

Reply:
SwitchErrorHandler does it as well

Reply:
for flows yes, but for main connection I do reconnect